execution-engine/src/wasi/strace.rs

```diff
@@ -9,30 +9,99 @@
 //! See the `LICENSE_MIT.markdown` file in the Veracruz root directory for
 //! information on licensing and copyright.
 
-#![allow(unused_variables)] // while most of the functions are unimplemented
-
 use super::common::MemoryHandler;
 use crate::fs::FileSystemResult;
 use std::fmt;
 
-pub struct Strace {}
+enum TraceState {
+    Func,
+    Args,
+    Result,
+}
+
+pub struct Strace {
+    enabled: bool,
+    state: TraceState,
+}
+
+// Convert a vector of bytes into a printable ASCII string.
+// The string may contain readable text, which is useful to display,
+// but it may also contain binary data, so we cannot decode it as UTF-8.
+fn strace_string(bytes: &[u8], max: usize) -> String {
+    let mut res = String::from("\"");
+    let n = if bytes.len() > max { max } else { bytes.len() };
+    for i in 0..n {
+        if 0x20 <= bytes[i] && bytes[i] < 0x7f {
+            if bytes[i] == b'\\' || bytes[i] == b'"' {
+                res.push_str("\\");
+            }
+            res.push_str(&String::from_utf8_lossy(&bytes[i..i + 1].to_vec()))
+        } else if bytes[i] == 9 {
+            res.push_str("\\t")
+        } else if bytes[i] == 10 {
+            res.push_str("\\n")
+        } else if i + 1 < n && b'0' <= bytes[i + 1] && bytes[i + 1] <= b'9' {
+            // The following character is a digit, so use three octal digits.
+            res.push_str(&format!("\\{:03o}", bytes[i]))
+        } else {
+            res.push_str(&format!("\\{:o}", bytes[i]))
+        }
+    }
+    res.push_str("\"");
+    if bytes.len() > max {
+        res.push_str("...")
+    }
+    res
+}
 
 impl Strace {
     pub fn func(enabled: bool, name: &str) -> Self {
-        // NOT YET IMPLEMENTED
-        Strace {}
+        if enabled {
+            eprint!("{}(", name)
+        };
+        Strace {
+            enabled,
+            state: TraceState::Func,
+        }
+    }
+
+    fn arg(&mut self) -> bool {
```

Comment:
This `returns` true when an argument is *not* expected, given the current state of the `Strace`, right?  I wonder if renaming this function to something more suggestive (e.g., `is_arg_expected()`) and introducing a special-purpose two-element type

```rust
enum ArgExpected {
  Expected,
  NotExpected
}
```

may not be a good idea, here?

All the checks at the beginning of functions, of the form:

```rust
if self.arg() {
    return;
}
```

then become something like:

```rust
if self.is_arg_expected().is_not_expected() {
  return;
}
```

where `is_not_expected()` is defined on `ArgExpected` in the obvious way...

Reply:
This was a hack to make the code more concise. I've separated the function from the test.